README.md
<!-- ix-docs-ignore -->
![imgix logo](https://assets.imgix.net/sdk-imgix-logo.svg)

`imgix-php` is a client library for generating image URLs with [imgix](https://www.imgix.com/). It is tested under PHP versions `5.6`, `7.0`, `7.1`, and `7.2`

[![Version](https://img.shields.io/packagist/v/imgix/imgix-php.svg)](https://packagist.org/packages/imgix/imgix-php)
[![Build Status](https://travis-ci.com/imgix/imgix-php.svg?branch=main)](https://travis-ci.com/imgix/imgix-php)
[![Downloads](https://img.shields.io/packagist/dt/imgix/imgix-php)](https://packagist.org/packages/imgix/imgix-php)
[![License](https://img.shields.io/github/license/imgix/imgix-php)](https://github.com/imgix/imgix-php/blob/main/LICENSE)
[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Fimgix%2Fimgix-php.svg?type=shield)](https://app.fossa.com/projects/git%2Bgithub.com%2Fimgix%2Fimgix-php?ref=badge_shield)

---
<!-- /ix-docs-ignore -->

- [Installation](#installation)
    * [Using Composer](#using-composer)
    * [Standalone](#standalone)
- [Usage](#usage)
- [Signed URLs](#signed-urls)
- [Srcset Generation](#srcset-generation)
    * [Fixed-Width Images](#fixed-width-images)
        + [Variable Quality](#variable-quality)
    * [Fluid-Width Images](#fluid-width-images)
        + [Custom Widths](#custom-widths)
        + [Width Ranges](#width-ranges)
        + [Width Tolerance](#width-tolerance)
- [The `ixlib` Parameter](#the-ixlib-parameter)
- [License](#license)

## Installation

### Using Composer

Define the following requirement in your `composer.json` file:

```json
{
  "require": {
    "imgix/imgix-php": "dev-main"
  }
}
```

And include the global `vendor/autoload.php` autoloader.

### Standalone

Just copy the files to your project, and include the `src/autoload.php` file. We recommend using Composer if at all possible.

## Usage

To begin creating imgix URLs programmatically, add the php files to your project (an example autoloader is also provided). The URL builder can be reused to create URLs for any
images on the domains it is provided.

```php
use Imgix\UrlBuilder;

$builder = new UrlBuilder("demos.imgix.net");
$params = array("w" => 100, "h" => 100);
echo $builder->createURL("bridge.png", $params);
// 'https://demos.imgix.net/bridge.png?h=100&w=100'
```

HTTPS support is available _by default_. However, if you need HTTP support, call `setUseHttps` on the builder:

```php
use Imgix\UrlBuilder;

$builder = new UrlBuilder("demos.imgix.net");
$builder->setUseHttps(false);
$params = array("w" => 100, "h" => 100);
echo $builder->createURL("bridge.png", $params);
// 'http://demos.imgix.net/bridge.png?h=100&w=100'
```

## Signed URLs

To produce a signed URL, you must enable secure URLs on your source and then
provide your signature key to the URL builder.

```php
use Imgix\UrlBuilder;

$builder = new UrlBuilder("demos.imgix.net");
$builder->setSignKey("test1234");
$params = array("w" => 100, "h" => 100);
echo $builder->createURL("bridge.png", $params);
// 'http://demos.imgix.net/bridge.png?h=100&w=100&s=bb8f3a2ab832e35997456823272103a4'
```

## Srcset Generation

The imgix-php package allows for generation of custom srcset attributes, which can be invoked through the `createSrcSet` method. By default, the generated srcset will allow for responsive size switching by building a list of image-width mappings.

```php
$builder = new UrlBuilder("demos.imgix.net", true, "my-key", false);
echo $builder->createSrcSet("image.png");
```

The above will produce the following srcset attribute value which can then be served to the client: 

``` html
https://demos.imgix.net/image.png?w=100&s=e415797545a77a9d2842dedcfe539c9a 100w,
https://demos.imgix.net/image.png?w=116&s=b2da46f5c23ef13d5da30f0a4545f33f 116w,
https://demos.imgix.net/image.png?w=135&s=b61422dead929f893c04b8ff839bb088 135w,
                                        ...
https://demos.imgix.net/image.png?w=7401&s=ad671301ed4663c3ce6e84cb646acb96 7401w,
https://demos.imgix.net/image.png?w=8192&s=a0fed46e2bbcc70ded13dc629aee5398 8192w
```

### Fixed-Width Images

In cases where enough information is provided about an image's dimensions, `createSrcSet` will instead build a srcset that will allow for an image to be served at different resolutions. The parameters taken into consideration when determining if an image is fixed-width are `w` and `h`.

By invoking `createSrcSet` with either a width **or** height provided, a different srcset will be generated for a fixed-width image instead.

```php
$builder = new UrlBuilder("demos.imgix.net", true, "my-key", false);
echo $builder->createSrcSet("image.png", array("h"=>800, "ar"=>"3:2", "fit"=>"crop"));
```

Will produce the following attribute value:

``` html
https://demos.imgix.net/image.png?ar=3%3A2&dpr=1&fit=crop&h=800&s=6cf5c443d1eb98bc3d96ea569fcef088 1x,
https://demos.imgix.net/image.png?ar=3%3A2&dpr=2&fit=crop&h=800&s=d60a61a5f34545922bd8dff4e53a0555 2x,
https://demos.imgix.net/image.png?ar=3%3A2&dpr=3&fit=crop&h=800&s=590f96aa426f8589eb7e449ebbeb66e7 3x,
https://demos.imgix.net/image.png?ar=3%3A2&dpr=4&fit=crop&h=800&s=c89c2fd3148957647e86cfc32ba20517 4x,
https://demos.imgix.net/image.png?ar=3%3A2&dpr=5&fit=crop&h=800&s=3d73af69d78d49eef0f81b4b5d718a2c 5x
```

For more information to better understand srcset, we highly recommend
[Eric Portis' "Srcset and sizes" article](https://ericportis.com/posts/2014/srcset-sizes/) which goes into depth about the subject.

#### Variable Quality

This library will automatically append a variable `q` parameter mapped to each `dpr` parameter when generating a [fixed-width image](#fixed-width-images) srcset. This technique is commonly used to compensate for the increased file size of high-DPR images.

Since high-DPR images are displayed at a higher pixel density on devices, image quality can be lowered to reduce overall file size––without sacrificing perceived visual quality. For more information and examples of this technique in action, see [this blog post](https://blog.imgix.com/2016/03/30/dpr-quality).

This behavior will respect any overriding `q` value passed in as a parameter. Additionally, it can be disabled altogether by passing `$disableVariableQuality = true` to `createSrcSet()`'s `$options`.

This behavior specifically occurs when a [fixed-width image](#fixed-width-images) is rendered, for example:

```php
// Note that `params=array("w" => 100)` allows `createSrcSet` to _infer_ the creation
// of a DPR based srcset attribute for fixed-width images.
$builder = new UrlBuilder("demos.imgix.net", true, "", false);
$params = array("w" => 100);
$srcset = $builder->createSrcSet($path="image.jpg", $params=$params);
```

The above will generate a srcset with the following `q` to `dpr` query `params`:

```html
https://demos.imgix.net/image.jpg?dpr=1&q=75&w=100 1x,
https://demos.imgix.net/image.jpg?dpr=2&q=50&w=100 2x,
https://demos.imgix.net/image.jpg?dpr=3&q=35&w=100 3x,
https://demos.imgix.net/image.jpg?dpr=4&q=23&w=100 4x,
https://demos.imgix.net/image.jpg?dpr=5&q=20&w=100 5x'
```

### Fluid-Width Images

#### Custom Widths

In situations where specific widths are desired when generating `srcset` pairs, a user can specify them by passing an array of positive integers as `'widths'` within the `$options` array:

``` php
$builder = new UrlBuilder("demos.imgix.net", true, "", false);
$opts = array('widths' => array(144, 240, 320, 446, 640));
$srcset = $builder->createSrcSet($path="image.jpg", $params=array(), $options=$opts);
```

```html
https://demos.imgix.net/image.jpg?w=144 144w,
https://demos.imgix.net/image.jpg?w=240 240w,
https://demos.imgix.net/image.jpg?w=320 320w,
https://demos.imgix.net/image.jpg?w=446 446w,
https://demos.imgix.net/image.jpg?w=640 640w
```

**Note**: in situations where a `srcset` is being rendered as a [fixed-width](#fixed-width-images) srcset, any custom `widths` passed in will be ignored.

Additionally, if both `widths` and a width `tol`erance are passed to the `createSrcSet` method, the custom widths list will take precedence.

#### Width Ranges

In certain circumstances, you may want to limit the minimum or maximum value of the non-fixed `srcset` generated by the `createSrcSet` method. To do this, you can specify the widths at which a srcset should `start` and `stop`:

```php
$builder = new UrlBuilder("demo.imgix.net", true, "", false);
$opts = array('start' => 500, 'stop' => 2000);
$srcset = $builder->createSrcSet($path="image.jpg", $params=array(), $options=$opts);
```

Formatted version of the above srcset attribute:

``` html
https://demo.imgix.net/image.jpg?w=500 500w,
https://demo.imgix.net/image.jpg?w=580 580w,
https://demo.imgix.net/image.jpg?w=673 673w,
https://demo.imgix.net/image.jpg?w=780 780w,
https://demo.imgix.net/image.jpg?w=905 905w,
https://demo.imgix.net/image.jpg?w=1050 1050w,
https://demo.imgix.net/image.jpg?w=1218 1218w,
https://demo.imgix.net/image.jpg?w=1413 1413w,
https://demo.imgix.net/image.jpg?w=1639 1639w,
https://demo.imgix.net/image.jpg?w=1901 1901w,
https://demo.imgix.net/image.jpg?w=2000 2000w
```

#### Width Tolerance

The `srcset` width `tol`erance dictates the maximum `tol`erated difference between an image's downloaded size and its rendered size.

For example, setting this value to `10` means that an image will not render more than 10% larger or smaller than its native size. In practice, the image URLs generated for a width-based srcset attribute will grow by twice this rate.

A lower tolerance means images will render closer to their native size (thereby increasing perceived image quality), but a large srcset list will be generated and consequently users may experience lower rates of cache-hit for pre-rendered images on your site.

By default, srcset width `tol`erance is set to 8 percent, which we consider to be the ideal rate for maximizing cache hits without sacrificing visual quality. Users can specify their own width tolerance by providing a positive scalar value as width `tol`erance:

```php
$builder = new UrlBuilder("demo.imgix.net", true, "", false);
$opts = array('start' => 100, 'stop' => 384, 'tol' => 0.20);
$srcset = $builder->createSrcSet($path="image.jpg", $params=array(), $options=$opts);
```

In this case, the width `tol`erance is set to 20 percent, which will be reflected in the difference between subsequent widths in a srcset pair:

```html
https://demo.imgix.net/image.jpg?w=100 100w,
https://demo.imgix.net/image.jpg?w=140 140w,
https://demo.imgix.net/image.jpg?w=196 196w,
https://demo.imgix.net/image.jpg?w=274 274w,
https://demo.imgix.net/image.jpg?w=384 384w
```

## The `ixlib` Parameter

For security and diagnostic purposes, we sign all requests with the language and version of library used to generate the URL.

This can be disabled by setting `setIncludeLibraryParam` to `False` like so:

``` php
$builder = new UrlBuilder("demo.imgix.net", true, "", false);
// Or by calling `setIncludeLibraryParam`
$builder->setIncludeLibraryParam(false);
```

## License
[![FOSSA Status](https://app.fossa.com/api/projects/git%2Bgithub.com%2Fimgix%2Fimgix-php.svg?type=large)](https://app.fossa.com/projects/git%2Bgithub.com%2Fimgix%2Fimgix-php?ref=badge_large)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="3.3.1"></a>

## [3.3.1](https://github.com/imgix/imgix-php/compare/3.3.0...3.3.1) (2021-03-24)
* docs: update fixed-widths section  ([#77](https://github.com/imgix/imgix-php/pull/77))
* fix: dpr srcset when only h param  ([#76](https://github.com/imgix/imgix-php/pull/76))
* docs: update travis badge to travis-ci.com  ([#74](https://github.com/imgix/imgix-php/pull/74))
* docs: reorder install instruction to promote usage of composer ([#72](https://github.com/imgix/imgix-php/pull/72))

<a name="3.3.0"></a>

## [3.3.0](https://github.com/imgix/imgix-php/compare/3.2.0...3.3.0) (2020-06-05)
* fix: normalize behavior of target widths ([#56](https://github.com/imgix/imgix-php/pull/56))
* fix: remove ensure even requirement ([#57](https://github.com/imgix/imgix-php/pull/57))
* feat: create custom srcsets ([#58](https://github.com/imgix/imgix-php/pull/58))
* feat: validate custom srcsets ([#61](https://github.com/imgix/imgix-php/pull/61))

<a name="3.2.0"></a>

## [3.2.0](https://github.com/imgix/imgix-php/compare/3.1.0...3.2.0) (2020-03-31)

* feat: use https by default ([#53](https://github.com/imgix/imgix-php/pull/53))

<a name="3.1.0"></a>

## [3.1.0](https://github.com/imgix/imgix-php/compare/3.0.0...3.1.0) (2019-08-29)

* feat: add srcset generation ([#47](https://github.com/imgix/imgix-php/pull/47))

<a name="3.0.0"></a>

## [3.0.0](https://github.com/imgix/imgix-php/compare/2.3.0...3.0.0) (2019-06-11)

* fix: remove deprecated domain sharding functionality ([#45](https://github.com/imgix/imgix-php/pull/45))

<a name="2.3.0"></a>

## [2.3.0](https://github.com/imgix/imgix-php/compare/2.2.0...2.3.0) (2019-05-06)

* deprecate domain sharding ([#42](https://github.com/imgix/imgix-php/pull/42)) ([#43](https://github.com/imgix/imgix-php/pull/43))

<a name="2.2.0"></a>

## [2.2.0](https://github.com/imgix/imgix-php/compare/2.1.1...2.2.0) (2019-04-08)

### Features

* add support for multiple URL parameter values by flattening nested arrays passed into $params ([#40](https://github.com/imgix/imgix-php/pull/40))

### Bug Fixes

* replace deprecated phpunit annotations with exception methods ([#39](https://github.com/imgix/imgix-php/pull/39))
* add domain validation at UrlBuilder initialization ([#41](https://github.com/imgix/imgix-php/pull/41)), fixes [#10](https://github.com/imgix/imgix-php/issues/10)


tests/tests.php
<?php

use Imgix\UrlBuilder;
use Imgix\UrlHelper;
use Imgix\ShardStrategy;

class ImgixTest extends PHPUnit_Framework_TestCase {

     /**
     * @expectedException        InvalidArgumentException
     * @expectedExceptionMessage UrlBuilder requires at least one domain
     */
    public function testURLBuilderRaisesExceptionOnNoDomains() {
        $domains = array();
        $ub = new URLBuilder($domains);
    }

    public function testHelperBuildSignedURLWithHashMapParams() {
        $params = array("w" => 500);
        $uh = new URLHelper("securejackangers.imgix.net", "chester.png", "http", "Q61NvXIy", $params);

        $this->assertEquals($uh->getURL(), "http://securejackangers.imgix.net/chester.png?w=500&s=0ddf97bf1a266a1da6c30c6ce327f917");
    }

    public function testHelperBuildSignedURLWithHashMapAndNoParams() {
        $params = array();
        $uh = new URLHelper("securejackangers.imgix.net", "chester.png", "http", "Q61NvXIy", $params);

        $this->assertEquals($uh->getURL(), "http://securejackangers.imgix.net/chester.png?s=cff7bdfd1b32d82e6b516f7fd3b4f1f4");
    }

    public function testHelperBuildSignedURLWithHashSetterParams() {
        $uh = new URLHelper("securejackangers.imgix.net", "chester.png", "http", "Q61NvXIy");
        $uh->setParameter("w", 500);
        $this->assertEquals($uh->getURL(), "http://securejackangers.imgix.net/chester.png?w=500&s=0ddf97bf1a266a1da6c30c6ce327f917");
    }

    public function testHelperBuildSignedURLWithHashSetterParamsHttps() {
        $uh = new URLHelper("securejackangers.imgix.net", "chester.png", "https", "Q61NvXIy");
        $uh->setParameter("w", 500);
        $this->assertEquals($uh->getURL(), "https://securejackangers.imgix.net/chester.png?w=500&s=0ddf97bf1a266a1da6c30c6ce327f917");
    }

    public function testUrlBuilderCycleShard() {
        // generate a url for the number of domains in use ensure they're cycled through...

        $domains = array("jackangers.imgix.net", "jackangers2.imgix.net", "jackangers3.imgix.net");

        $ub = new URLBuilder($domains);
        $ub->setShardStrategy(ShardStrategy::CYCLE);

        for ($i = 0; $i < 100; $i++) {
			$used = array();
			foreach ($domains as $domain) {
				$url = $ub->createURL("chester.png");
				$curDomain = parse_url($url)["host"];
				$this->assertFalse(in_array($curDomain, $used));
				$used[] = $curDomain;
			}
        }
	}

	public function testUrlBuilderCRCShard() {

        $domains = array("jackangers.imgix.net", "jackangers2.imgix.net", "jackangers3.imgix.net");

        $ub = new URLBuilder($domains);

        $paths = array("chester.png", "chester1.png", "chester2.png");

        foreach ($paths as $path) {
            $testDomain = parse_url($ub->createURL($path))['host'];

            for ($i = 0; $i < 20; $i++) {
                $this->assertEquals($testDomain, parse_url($ub->createURL($path))['host']);
            }
        }
	}

    public function testExamplePlain() {
        $builder = new UrlBuilder("demos.imgix.net");

        $params = array("w" => 100, "h" => 100);
        $url = $builder->createURL("bridge.png", $params);

        $this->assertEquals($url, "http://demos.imgix.net/bridge.png?h=100&w=100");
    }

    public function testExamplePlainHttps() {
        $builder = new UrlBuilder("demos.imgix.net");

        $builder->setUseHttps(true);
        $params = array("w" => 100, "h" => 100);
        $url = $builder->createURL("bridge.png", $params);

        $this->assertEquals($url, "https://demos.imgix.net/bridge.png?h=100&w=100");
    }

    public function testExamplePlainSecure() {
        $builder = new UrlBuilder("demos.imgix.net");
        $builder->setSignKey("test1234");
        $params = array("w" => 100, "h" => 100);
        $url = $builder->createURL("bridge.png", $params);

        $this->assertEquals($url, "http://demos.imgix.net/bridge.png?h=100&w=100&s=bb8f3a2ab832e35997456823272103a4");
    }


}

src/Imgix/UrlHelper.php
<?php

namespace Imgix;

class UrlHelper {

    private $domain;
    private $path;
    private $scheme;
    private $signKey;
    private $params;

    public function __construct($domain, $path, $scheme = "http", $signKey = "", $params = array()) {
        $this->domain = $domain;
        $this->path = $this->formatPath($path);
        $this->scheme = $scheme;
        $this->signKey = $signKey;
        $this->params = $params;
    }

    public function formatPath($path) {
        if (!is_string($path) || strlen($path) < 1)
            return '/';

        // Strip leading slash first (we'll re-add after encoding)
        $path = preg_replace("/^\//", "", $path);

        if (preg_match("/^https?:\/\//", $path)) {
            // If this path is a full URL, encode the entire thing
            $path = rawurlencode($path);
        } else if (preg_match("/^https?:\/\/[^\s\/$.?#]*\.[^\s]*$/", rawurldecode($path))) {
            // Using @stephenhay's solution from https://mathiasbynens.be/demo/url-regex
            // to ensure URL's validity.
            // $path looks like a valid encoded URL, however, it may still have
            // unencoded unicode characters.
            $path = preg_replace_callback("/([^\w\-\/\:@%])/", function ($match) {
                return rawurlencode($match[0]);
            }, $path);
        } else {
            // If this path is just a path, only encode certain characters
            $path = preg_replace_callback("/([^\w\-\/\:@])/", function ($match) {
                return rawurlencode($match[0]);
            }, $path);
        }

        // Add a leading slash before returning
        return '/' . $path;
    }

    public function setParameter($key, $value) {
        if ($key && ($value || $value === 0)) {
            $this->params[$key] = $value;
        } else {
            if (array_key_exists($key, $this->params)) {
                unset($this->params[$key]);
            }
        }
    }

    public function deleteParameter($key) {
        unset($this->params[$key]);
    }

    public function getURL() {
        $queryPairs = array();

        if ($this->params) {
            ksort($this->params);

            foreach ($this->params as $key => $val) {
                if (substr($key, -2) == '64') {
                    $encodedVal = self::base64url_encode($val);
                } else {
                    $encodedVal = is_array($val) ? rawurlencode(implode(',',$val)) : rawurlencode($val);
                }

                $queryPairs[] = rawurlencode($key) . "=" . $encodedVal;
            }
        }

        $query = join("&", $queryPairs);
        if ($query) {
            $query = '?' . $query;
        }

        if ($this->signKey) {
            $toSign = $this->signKey . $this->path . $query;
            $sig = md5($toSign);
            if ($query) {
                $query .= "&s=" . $sig;
            } else {
                $query = "?s=" . $sig;
            }
        }

        $url_parts = array('scheme' => $this->scheme, 'host' => $this->domain, 'path' => $this->path, 'query' => $query);

        return self::joinURL($url_parts);
    }

    private static function base64url_encode($data) {
        return rtrim(strtr(base64_encode($data), '+/', '-_'), '=');
    }

    private static function joinURL($parts) {
        $url = $parts['scheme'] . '://' . $parts['host'] . $parts['path'] . $parts['query'];

        return $url;
    }
}


src/Imgix/UrlBuilder.php
<?php

namespace Imgix;

use Imgix\Validator;

class UrlBuilder {

    private $currentVersion = "3.3.1";
    private $domain;
    private $useHttps;
    private $signKey;

    const TARGET_WIDTHS = array(
        100, 116, 134, 156, 182, 210, 244, 282,
        328, 380, 442, 512, 594, 688, 798, 926,
        1074, 1246, 1446, 1678, 1946, 2258, 2618,
        3038, 3524, 4088, 4742, 5500, 6380, 7400, 8192);

    // define class constants
    // should be private; but visibility modifiers are not supported php version <7.1
    const TARGET_RATIOS = array(1, 2, 3, 4, 5);

    const DPR_QUALITIES = array(1 => 75, 2 => 50, 3 => 35, 4 => 23, 5 => 20);

    const MIN_WIDTH = 100;
    const MAX_WIDTH = 8192;
    const SRCSET_WIDTH_TOLERANCE = 0.08;

    // constants cannot be dynamically assigned; keeping as a class variable instead
    private $targetWidths;

    public function __construct($domain, $useHttps = true, $signKey = "", $includeLibraryParam = true) {

        if (!is_string($domain)) {
            throw new \InvalidArgumentException("UrlBuilder must be passed a string domain");
        }

        $this->domain = $domain;
        $this->validateDomain($this->domain);        

        $this->useHttps = $useHttps;
        $this->signKey = $signKey;
        $this->includeLibraryParam = $includeLibraryParam;
        $this->targetWidths = $this->targetWidths();
    }

    private function validateDomain($domain) {
        $DOMAIN_PATTERN = "/^(?:[a-z\d\-_]{1,62}\.){0,125}(?:[a-z\d](?:\-(?=\-*[a-z\d])|[a-z]|\d){0,62}\.)[a-z\d]{1,63}$/";

        if(!preg_match($DOMAIN_PATTERN, $domain)) {
            throw new \InvalidArgumentException('Domain must be passed in as fully-qualified ' . 
            'domain name and should not include a protocol or any path element, i.e. ' .
            '"example.imgix.net".'); 
        }
    }

    public function setSignKey($key) {
        $this->signKey = $key;
    }

    public function setUseHttps($useHttps) {
        $this->useHttps = $useHttps;
    }

    public function setIncludeLibraryParam($includeLibraryParam) {
        $this->includeLibraryParam = $includeLibraryParam;
    }

    public function createURL($path, $params=array()) {
        $scheme = $this->useHttps ? "https" : "http";
        $domain = $this->domain;

        if ($this->includeLibraryParam) {
            $params['ixlib'] = "php-" . $this->currentVersion;
        }

        $uh = new UrlHelper($domain, $path, $scheme, $this->signKey, $params);

        return $uh->getURL();
    }

    public function createSrcSet($path, $params=array(), $options=array()) {
        $widthsArray = isset($options['widths']) ? $options['widths'] : NULL;

        if (isset($widthsArray)) {
            Validator::validateWidths($widthsArray);
            return $this->createSrcSetPairs($path, $params=$params, $widthsArray);
        }
        
        $_start = isset($options['start']) ? $options['start'] : self::MIN_WIDTH;
        $_stop = isset($options['stop']) ? $options['stop'] : self::MAX_WIDTH;
        $_tol = isset($options['tol']) ? $options['tol'] : self::SRCSET_WIDTH_TOLERANCE;
        $_disableVariableQuality = isset($options['disableVariableQuality'])
                ? $options['disableVariableQuality'] : false;

        if ($this->isDpr($params)) {
            return $this->createDPRSrcSet(
                $path=$path, 
                $params=$params, 
                $disableVariableQuality=$_disableVariableQuality);
        }
        else {
            $targets = $this->targetWidths($start=$_start, $stop=$_stop, $tol=$_tol);
            return $this->createSrcSetPairs($path, $params=$params, $targets=$targets);
        }
    }

    /**
     * Generate a list of target widths.
     * 
     * This function generates an array of target widths used to
     * width-describe image candidate strings (URLs) within a
     * srcset attribute.
     * 
     * This function returns an array of integer values that denote
     * image target widths. This array begins with `$start`
     * and ends with `$stop`. The `$tol` or tolerance value dictates
     * the amount of tolerable width variation between each width
     * in the range of values that lie between `$start` and `$stop`.
     * 
     * @param  int $start Starting minimum width value.
     * @param  int $stop Stopping maximum width value.
     * @param  int $tol Tolerable amount of width variation.
     * @return int[] $resolutions An array of integer values.
     */
    public function targetWidths(
        $start=self::MIN_WIDTH,
        $stop=self::MAX_WIDTH,
        $tol=self::SRCSET_WIDTH_TOLERANCE) {

        if ($start == $stop) {
            return array((int) $start);
        }

        Validator::validateMinMaxTol($start, $stop, $tol);
        $resolutions = array();

        while ($start < $stop && $start < self::MAX_WIDTH) {
            array_push($resolutions, (int) round($start));
            $start *= 1 + $tol * 2;
        }

        // The most recently appended value may, or may not, be
        // the `stop` value. In order to be inclusive of the
        // stop value, check for this case and add it, if necessary.
        if (end($resolutions) < $stop) {
            array_push($resolutions, (int) $stop);
        }
        
        return $resolutions;
    }

    private function isDpr($params) {
        if (empty($params)) {
            // If the params array is empty, then
            // it is _not_ dpr based.
            return false;
        }

        $hasWidth = array_key_exists('w', $params) ? $params['w'] : NULL;
        $hasHeight = array_key_exists('h', $params) ? $params['h'] : NULL;
        $hasAspectRatio = array_key_exists('ar', $params) ? $params['ar'] : NULL;

        // If `params` have a width or height parameter then the
        // srcset to be constructed with these params _is dpr based
        return $hasWidth || $hasHeight;
    }

    private function createDPRSrcSet($path, $params, $disableVariableQuality=false) {
        $srcset = "";

        $size = count(self::TARGET_RATIOS);
        for ($i = 0; $i < $size; $i++) {
            $currentParams = $params;
            $currentParams['dpr'] = $i+1;
            $currentRatio = self::TARGET_RATIOS[$i];
            // If variable quality output has been disabled _and_
            // the `q` param _has not_ been passed:
            if (!$disableVariableQuality && !isset($params["q"])) {
                $currentParams["q"] = self::DPR_QUALITIES[$i + 1];
            }
            $srcset .= $this->createURL($path, $currentParams) . " " . $currentRatio . "x,\n";
        }

        return substr($srcset, 0, strlen($srcset) - 2);
    }

    private function createSrcSetPairs($path, $params, $targets=self::TARGET_WIDTHS) {
        $srcset = "";
        $currentWidth = NULL;
        $currentParams = NULL;

        $size = count($targets);
        for ($i = 0; $i < $size; $i++) {
            $currentWidth = $targets[$i];
            $currentParams = $params;
            $currentParams['w'] = $currentWidth;
            $srcset .= $this->createURL($path, $currentParams) . " " . $currentWidth . "w,\n";
        }

        return substr($srcset, 0, strlen($srcset) - 2);
    }
}
